SRE Operation Trails (#42)
quipper/quipper.github.io: _posts/2019-05-21-sre-operation-trails.markdown

title:  "SRE Operation Trails"
date:   2019-05-21 00:00:00
author: rbmrclo
comments: true
---

## Intro

Hello! This is [@rbmrclo](https://github.com/rbmrclo) from Site Reliability Engineering team.
Today, let me share about "Operation Trails" which is an important part of our workflow when performing tasks that involves manual operation.

## Background

In the SRE team, we have a 50/50 golden rule for how we manage our time every day.

To summarise, half of our day usually goes to **"proactive"** tasks, which are generally the main projects that contributes to our growth as a diverse tech team (we usually have a roadmap for this).
The rest of our time are spent on **"reactive"** tasks, which are essential to maintain the stability and reliability of our services, as well as to keep the development speed across each teams.

It can be visualised in blocks like this:

![sre-tasks](/assets/article_images/2019-05-21-sre-operation-trails/proactive-reactive-sre.png)

In this article, I will be focusing on our **reactive tasks** and explain in detail how we manage to work seamlessly within our team and avoid "_mottainai_". (I’ll be explaining more on this later)

## Reactive tasks

> As a global company, each SRE members are attending to the needs of multiple teams, and in multiple time zones. This means that each members are also working on their own pace.

Reviewer: As a global company, each SRE member attends to the needs of multiple teams in different timezones. This also means that each member is working at their own pace.

> As a global company, each SRE members are attending to the needs of multiple teams, and in multiple time zones. This means that each members are also working on their own pace.

Reviewer: ```suggestion
As a global company, each SRE member attends to the needs of multiple teams in different timezones. This also means that each member is working at their own pace.
```

---
layout: post
title:  "SRE Operation Trails"
date:   2019-05-21 00:00:00
author: rbmrclo
comments: true
---

## Intro

Hello! This is [@rbmrclo](https://github.com/rbmrclo) from Site Reliability Engineering team.

> Today, let me share about **"Operation Trails"** (a term we use in our team) which is an important part of our workflow when performing tasks that involves manual operation.

Reviewer: ```suggestion
Today, let me share about **"Operation Trails"** (a term we use in our team) which is an important part of our workflow when performing tasks that involve manual operation.
```

## Background

In the SRE team, we have a 50/50 rule for how we manage our time every day.

> To summarise, half of our day usually goes to **proactive** tasks, which are generally the main projects that contributes to our growth as a diverse tech team (we usually have a roadmap for this).

Reviewer: ```suggestion
To summarise, half of our day usually goes to **proactive** tasks which are generally the main projects that contribute to our growth as a diverse tech team (we usually have a roadmap for this).
```

> The rest of our time are spent on **reactive** tasks, which are essential to maintain the stability and reliability of our services, as well as to keep the development speed stable across each teams.

Reviewer: ```suggestion
The rest of our time is spent on **reactive** tasks which are essential to maintain the stability and reliability of our services, as well as to keep the development speed stable across each team.
```

It can be visualised in blocks like this:

![sre-tasks](/assets/article_images/2019-05-21-sre-operation-trails/proactive-reactive-sre.png)

In this article, I will be focusing on our **reactive tasks** and explain in detail how we manage to work seamlessly within our team and avoid _mottainai_ (I’ll be explaining this later).

## Daily Situation

As a global company, each SRE member attends to the needs of multiple teams in different timezones. This also means that each member is working at their own pace.

> Some members might be working on a normal routine today with their proactive tasks; some will be performing a maintenance task tonight (midnight!); or some are already attending a service outage incident while I'm writing this blog post!

Reviewer: ```suggestion
Some members might be working on a normal routine today with their proactive tasks; some will be performing a maintenance task tonight (midnight!); and some might already be attending to a service outage incident while I'm writing this blog post!
```

Let's illustrate that again with my favorite blocks.

![sre-isolated](/assets/article_images/2019-05-21-sre-operation-trails/sre-isolated.png)

> My point here is that most of the time, each of us are working in isolated manner. However, there's one exception, and this is when **Operation Trails** comes in.

Reviewer: ```suggestion
My point here is that most of the time, each of us is working in an isolated manner. However, there's one exception and this is when **Operation Trails** comes in.
```

## Operation Trails for Reactive Tasks

Imagine that you are working on a task, with your headphones on, enjoying your favorite bubble milk tea, listening to the playlist of Queen, in-the-zone and cannot be disturbed by humans.

Suddenly, an alert has been triggered for a specific monitor. Say the staging cluster died, hence, no developers could connect to the staging servers to test their newly implemented features - a major blocker!

**Call of duty.** Upon receiving the alert message, you quickly checked the issue and created an **Operation Trail**.

- First, you informed the other SRE team members that you are now checking the issue.
  - You are now considered as the assigned person. (ownership is part of our culture!)
  - This is also when the operation trail starts.

>   - All SRE members are now informed that someone is checking the issue, they are also watching the operation trail in parallel.

Reviewer: ```suggestion
  - All SRE members are now informed that someone is checking the issue. They are also watching the operation trail in parallel.
```

- Next, continuously post updates of what you're currently doing. (who did what when - like audit trails!)
  - While posting updates, other SRE team members could either give suggestions, join the ongoing operation, or just watch the trail. (it all depends on the severity of the situation)
- Lastly, you inform everyone when the task is finished or when the issue has been resolved. :tada:

#### Here's the bird's eye view of what happened.

![respond-alert](/assets/article_images/2019-05-21-sre-operation-trails/proactive-response.gif)

> **:memo: Every operations are in a single thread**

Reviewer: ```suggestion
**:memo: Every operation is in a single thread**
```

![operation-trail-example](/assets/article_images/2019-05-21-sre-operation-trails/operation-trail.png)

**:bell: Live reporting**

![actual-trail-example](/assets/article_images/2019-05-21-sre-operation-trails/actual-trail-1.png)

**:white_check_mark: Avoid operation conflicts by using call-to-actions**

![cta](/assets/article_images/2019-05-21-sre-operation-trails/cta.png)

## Summary

#### Slack Threads

> - In simplest term, operation trails are chat-based and it happens in real-time. We fully utilize slack threads for this.

Reviewer: ```suggestion
- In simple terms, operation trails are chat-based and happen real-time. We fully utilize slack threads for this.
```

- An SRE member can start an operation trail and resolve it by himself/herself, or another SRE member can join the trail to speed up resolving the task at hand.

#### Avoid [**Mottainai**](https://en.wikipedia.org/wiki/Mottainai) (もったいない)

> The term in Japanese conveys a sense of regret over waste; the exclamation "Mottainai!" can translate as "What a waste!"

- By establishing a live reporting culture in your team, you can eliminate waste of time.
  - For example, when an SRE member initiates that he/she is already responding to the issue, the other SRE members can just watch the trail while working on their current tasks normally. They don't need to pause as well, maximizing the use of their time.
- By actively posting updates in the operation trail, other members can provide relevant suggestions or possible solutions in order to speed up the operation.

#### Being a team-player

- Operation Trails improve the communication skill of an individual by being able to explain what's happening and what they are doing.
- As spectator of the trail, you can determine if the operation is going smoothly or a call for help is needed - evolving into a "pair operation".

> - It also improves harmony in the team since this is one of the times where all of us in SRE team can meet and collaborate with each other, given that we have individual tasks at hand.

Reviewer: ```suggestion
- It also improves harmony in the team since this is one of the times when all of us in SRE team can meet and collaborate with each other, given that we have individual tasks too.
```

---
layout: post
title:  "SRE Operation Trails"
date:   2019-05-21 00:00:00
author: rbmrclo
comments: true
---

## Intro

Hello! This is [@rbmrclo](https://github.com/rbmrclo) from Site Reliability Engineering team.
Today, let me share about "Operation Trails" which is an important part of our workflow when performing tasks that involves manual operation.

## Background

In the SRE team, we have a 50/50 rule for how we manage our time every day.

> To summarise, half of our day usually goes to **"proactive"** tasks, which are generally the main projects that contributes to our growth as a diverse tech team (we usually have a roadmap for this).

Reviewer: ```suggestion
The rest of our time is spent on **"reactive"** tasks which are essential to maintain the stability and reliability of our services, as well as to keep the development speed stable across each team.
```

> The rest of our time are spent on **"reactive"** tasks, which are essential to maintain the stability and reliability of our services, as well as to keep the development speed across each teams.

Reviewer: ```suggestion
The rest of our time is spent on **"reactive"** tasks which are essential to maintain the stability and reliability of our services, as well as to keep the development speed stable across each team.
```

It can be visualised in blocks like this:

![sre-tasks](/assets/article_images/2019-05-21-sre-operation-trails/proactive-reactive-sre.png)

> In this article, I will be focusing on our **reactive tasks** and explain in detail how we manage to work seamlessly within our team and avoid _mottainai_. (I’ll be explaining more on this later)

Reviewer: ```suggestion
In this article, I will be focusing on our **reactive tasks** and explain in detail how we manage to work seamlessly within our team and avoid _mottainai_ (I’ll be explaining this later).
```

## Reactive Tasks

As a global company, each SRE member attends to the needs of multiple teams in different timezones. This also means that each member is working at their own pace.

> Some members might be working on a normal routine today with their proactive tasks; some will be performing a maintenance task tonight (midnight!); or some are already attending a service outage incident while i'm writing this blog post!

Reviewer 2: s/i'm/I'm/

Let's illustrate that again with my favorite blocks.

![sre-isolated](/assets/article_images/2019-05-21-sre-operation-trails/sre-isolated.png)

My point here is that most of the time, each of us are working in isolated manner. However, there's one exception, and this is when **Operation Trails** comes in.

## Operation Trails

> Imagine that you are working on a task, with your headphones on, enjoying your favorite bubble milk tea, listening to the playlist of Queen, in-the-zone and cannot be disturbed by humans.

Reviewer 2: Queen 😎 Do you like them?

Author: Lol. Yes!

> Suddenly, an alert has been triggered for a specific monitor. Say the staging cluster is now dead, hence, no developers could connect to the staging servers and test their newly implemented features - a major blocker!

Reviewer: ```suggestion
Suddenly, an alert has been triggered for a specific monitor. Say the staging cluster died, hence, no developers could connect to the staging servers to test their newly implemented features - a major blocker!
```

**Call of duty.** Upon receiving the alert message, you quickly checked the issue and created an **Operation Trail**.

- First, you informed the other SRE team members that you are now checking the issue.

>   - You are now considered as the assigned person. (we have a proactive culture!)

Reviewer 2: >we have a proactive culture!

I feel a little confused with the context of the `proactive tasks` you mentioned earlier.
In fact, this response will be reactive.
Is there another word to indicate this culture?

Ownership? 🤔 sorry I don't find a word though 😅

Author: Makes sense! It's better to use "Ownership" since it's also a Quipper value! I'll change it.

- Lastly, you inform everyone when the task is finished or when the issue has been resolved. :tada:

#### Here's the bird's eye view of what happened.

![respond-alert](/assets/article_images/2019-05-21-sre-operation-trails/respond-alert.png)

**:memo: Every operations are in a single thread**

![operation-trail-example](/assets/article_images/2019-05-21-sre-operation-trails/operation-trail.png)

**:bell: Live reporting**

![actual-trail-example](/assets/article_images/2019-05-21-sre-operation-trails/actual-trail-1.png)

**:white_check_mark: Avoid operation conflicts by using call-to-actions**

![cta](/assets/article_images/2019-05-21-sre-operation-trails/cta.png)

## Summary

#### Slack Threads

- In simplest term, operation trails are chat-based and it happens in real-time. We fully utilize slack threads for this.
- An SRE member can start an operation trail and resolve it by himself/herself, or another SRE member can join the trail to speed up resolving the task at hand.

#### Avoid [**Mottainai**](https://en.wikipedia.org/wiki/Mottainai) (もったいない)

> > The term in Japanese conveys a sense of regret over waste; the exclamation "Mottainai!" can translate as "What a waste!"

Reviewer 2: That's right 😄

- By establishing a live reporting culture in your team, you can eliminate waste of time.
  - For example, when an SRE member initiates that he/she is already responding to the issue, the other SRE members can just watch the trail while working on their current tasks normally. They don't need to pause as well, maximizing the use of their time.
- By actively posting updates in the operation trail, other members can provide relevant suggestions or possible solutions in order to speed up the operation.

#### Being a team-player

> - Operation trails improves the communication skill of an individual by being able to explain what's happening and what they are doing.

Reviewer: ```suggestion
- Operation trails improve the communication skill of an individual by being able to explain what's happening and what they are doing.
```

- As spectator of the trail, you can determine if the operation is going smoothly or a call for help is needed - evolving into a "pair operation".
- It also improves harmony in the team since this is one of the times where all of us in SRE team can meet and collaborate with each other, given that we have individual tasks at hand.

#### Acknowledgements

> - There's also a [blog post in japanese](https://blog.kyanny.me/entry/2016/11/11/021955) which is the main inspiration of this post.

Reviewer 2: 👍 cc: @kyanny
